pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarClientImpl.java

```diff
@@ -990,8 +991,13 @@ public CompletableFuture<ClientCnx> getConnectionToServiceUrl() {
         return getConnection(address, address, cnxPool.genRandomKeyToSelectCon());
     }
 
-    public CompletableFuture<ClientCnx> getProxyConnection(final InetSocketAddress logicalAddress,
+    public CompletableFuture<ClientCnx> getProxyConnection(final URI redirectedClusterURI,
+                                                           final InetSocketAddress logicalAddress,
                                                            final int randomKeyForSelectConnection) {
+
+        LookupService lookup =
+                redirectedClusterURI == null ? this.lookup : getLookup(redirectedClusterURI.toString());
```

Comment:
I think that's a fair concern.

It appears that the connection cache key should be "logicalAddr + physicalAddr + random-key", and the pool can be a flat map, like the one below. I am not sure why physicalAddr is excluded now.

`ConcurrentHashMap<String, CompletableFuture<ClientCnx> pool>`

I wonder if we need to address this in a separate PR, as this is a design change for the connection cache. (We could isolate this PR to fix the wrong physical address after migration).


Also, the below code(and other getConnection funcs) might need to be revisited since they generate a new ran-key for every getConnection, which is a different connection cache access behavior after migration.

https://github.com/apache/pulsar/blob/master/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java#L115

Reply:
> It appears that the connection cache key should be "logicalAddr + physicalAddr + random-key", and the pool can be a flat map, like the one below. I am not sure why physicalAddr is excluded now.

The correct solution might not be so obvious. I think that the commit & PR history reveals some important details for understanding it. IIRC, the connection pool has some gotchas that easily trick into making wrong assumptions.

Reply:
At least for me, the connection pool and the various concepts around it in the Pulsar client are very misleading.

Reply:
Raised a PR to add the physical address in the key https://github.com/apache/pulsar/pull/22196

Reply:
https://github.com/apache/pulsar/pull/22196 got merged. Plz continue this PR review.